packages/beacon-node/src/chain/chain.ts

```diff
@@ -934,6 +937,42 @@ export class BeaconChain implements IBeaconChain {
     if (headState) {
       this.opPool.pruneAll(headBlock, headState);
     }
+
+    const cpEpoch = cp.epoch;
+    const eip6110Epoch = headState?.config.EIP6110_FORK_EPOCH ?? Infinity;
+
+    if (headState === null) {
+      this.logger.verbose("Head state is null");
+    } else if (cpEpoch >= eip6110Epoch) {
+      const headEpoch = headState.epochCtx.epoch;
+
+      // Get the validator.length from the state at cpEpoch
+      // We are confident the last element in the list is from headEpoch
+      // Thus we query from the end of the list. (cpEpoch - headEpoch - 1) is negative number
+      const pivotValidatorIndex = headState.epochCtx.historicalValidatorLengths.get(cpEpoch - headEpoch - 1);
```

Comment:
this part is quite confusing to me. Can you make `historicalValidatorLengths` private, add a new method like `getValidatorCountAtEpoch()` to `EpochCache`? Then unit test that method to make sure it's correct

with that the consumer code here becomes easier to understand and maintain